pkg/virt-handler/vm.go

```diff
@@ -2278,6 +2248,81 @@ func (d *VirtualMachineController) setVMIGuestTime(vmi *v1.VirtualMachineInstanc
 	return nil
 }
 
+func (d *VirtualMachineController) checkCPUForMigration(vmi *v1.VirtualMachineInstance) error {
+	if vmi.Spec.Domain.CPU == nil {
+		if d.clusterConfig.CPUMigrationCompatibilityEnabled() {
+			return fmt.Errorf("cannot migrate VMI with null CPU configuration")
+		}
+		return nil
+	}
+	model := vmi.Spec.Domain.CPU.Model
+	switch model {
+	case "", v1.CPUModeHostModel:
+		if d.clusterConfig.CPUMigrationCompatibilityEnabled() {
+			return fmt.Errorf("cannot migrate VMI with current CPU configuration: %s", model)
+		}
+	case v1.CPUModeHostPassthrough:
+		if !d.clusterConfig.MigrationOfHostPassthroughEnabled() {
+			return fmt.Errorf("cannot migrate VMI with current CPU configuration: %s", model)
+		}
+	}
+	return nil
+
+}
+
+func (d *VirtualMachineController) isMigratable(vmi *v1.VirtualMachineInstance) {
+	liveMigrationPossible := true
+
+	if err := d.checkCPUForMigration(vmi); err != nil {
+		liveMigrationCondition := v1.VirtualMachineInstanceCondition{
+			Type:    v1.VirtualMachineInstanceIsMigratable,
+			Status:  k8sv1.ConditionFalse,
+			Message: err.Error(),
+			Reason:  v1.VirtualMachineInstanceReasonCPUNotMigratable,
+		}
+		vmi.Status.Conditions = append(vmi.Status.Conditions, liveMigrationCondition)
+		liveMigrationPossible = false
+	}
+
+	isBlockMigration, err := d.checkVolumesForMigration(vmi)
+	if err != nil {
+		liveMigrationCondition := v1.VirtualMachineInstanceCondition{
+			Type:    v1.VirtualMachineInstanceIsMigratable,
+			Status:  k8sv1.ConditionFalse,
+			Message: err.Error(),
+			Reason:  v1.VirtualMachineInstanceReasonDisksNotMigratable,
+		}
+		vmi.Status.Conditions = append(vmi.Status.Conditions, liveMigrationCondition)
+		liveMigrationPossible = false
+	}
+
+	err = d.checkNetworkInterfacesForMigration(vmi)
+	if err != nil {
+		liveMigrationCondition := v1.VirtualMachineInstanceCondition{
+			Type:    v1.VirtualMachineInstanceIsMigratable,
+			Status:  k8sv1.ConditionFalse,
+			Message: err.Error(),
+			Reason:  v1.VirtualMachineInstanceReasonInterfaceNotMigratable,
+		}
+		vmi.Status.Conditions = append(vmi.Status.Conditions, liveMigrationCondition)
+		liveMigrationPossible = false
+	}
+
+	if liveMigrationPossible {
+		liveMigrationCondition := v1.VirtualMachineInstanceCondition{
+			Type:   v1.VirtualMachineInstanceIsMigratable,
+			Status: k8sv1.ConditionTrue,
+		}
+		vmi.Status.Conditions = append(vmi.Status.Conditions, liveMigrationCondition)
```

Comment:
Are we sure we want to pile these up? the code is looking for one `VirtualMachineInstanceIsMigratable` condition when we decide to migrate. I don't think it's useful.
These conditions are static and will stay with the VMI for its lifetime. Perhaps we could just expand the `Reason` field or maybe just reason is enough.

Reply:
@vladikr Wouldn't it be useful to know all the reasons why the VM is not live-migratable? It will make me as a user to solve all the issues at once instead of going back and forth. IIUC VMI lifetime will end anyway if an attempt to to make the VMI live-migratable will take place, thus the conditions will reset.

Reply:
@vladikr @fabiand I can concatenate the error and reason strings into a single string and produce a single condition. Will that be user-friendly enough?